docs/javascript-unit-tests.md

```diff
@@ -0,0 +1,712 @@
+
+#### Loading dependencies
+
+Line 1 of any spec file will use an `import` statement to
+include the JavaScript file that you are testing.
+Additional dependencies should be added in the same manner.
+
+```js
+import * as sample from '../../../../cfgov/unprocessed/js/modules/sample.js';
+```
+
+Some test files use `const` declarations to
+`require` scripts instead of `import`,
+because those files were written before `import` was available.
+We prefer to use `import` because it allows for
+[tree shaking in Webpack](https://webpack.js.org/guides/tree-shaking/),
+meaning if two modules are importing the same module
+it should only be included in the bundle once,
+whereas with `require` it would be included twice.
+
+A consequence is that variables can't be used in the import path,
+as they prevent Webpack from figuring out which modules are duplicates.
+For example, this snippet shows how a `require` statement
+should be converted to an `import` statement,
+but without including the `BASE_JS_PATH` variable in the file path:
+
+```js
+// works but could duplicate footer-button.js if other files also require it
+const FooterButton = require( BASE_JS_PATH + 'modules/footer-button.js' );
+
+// doesn't work and the build will fail
+import * as FooterButton from BASE_JS_PATH + 'modules/footer-button.js';
+
+// is ugly but it works and supports tree shaking
+import * as FooterButton from '../../../../cfgov/unprocessed/js/modules/footer-button.js';
+```
+
+`import` also provides a benefit in that you can choose specific parts
+of a module so that you only import the dependencies you need.
+For testing purposes, we will typically import the whole module
+to make sure we have full test coverage.
+[Read the `import` reference guide on MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import)
+on how to implement `import` for different use cases.
+
+#### The `describe` function
+
+In Jest (whose syntax is based on Jasmine),
+[`describe` blocks](https://jestjs.io/docs/en/api#describename-fn)
+serves as an organizational structures that you can use
+to outline the methods you need in your JS module.
+
+The root `describe` method is where we put
+the name of the JavaScript module we are testing.
+For the sample, the name of our module is `sample`,
+so we set this up on line 4 of `sample-spec.js`:
+
+```js
+describe( 'sample', () => {
+  …
+} );
+```
+
+This module name will appear in your test output in the console
+when the test is running:
+
+![Console output for failing unit test that reads 'sample: should return a string with expected value'](img/js-unit-test-output.png)
+
+More complex tests will have additional `describe` blocks –
+children of the root `describe` block –
+that should correspond with a particular method in the module.
+For example, if we want to add more functionality to our sample JS,
+we could start by writing these tests in `sample-spec.js`:
+
+```js
+describe( 'sample', () => {
```

Comment:
You might consider making `sample` a class as it becomes a bit clearer what the structure should be here (things in parens would be omitted in actual code):

```js
describe( 'Sample (class)', () => {
    describe( 'gimmeString() (method)', () => {
        it( 'should … (expected behavior)', () => {
    describe( 'anotherMethod()', () => {
        it( 'should … (expected behavior)', () => {
…
```

Reply:
Well… probably more complexity than you want to add to the example, but classes are guaranteed to have a structure of the class containing methods, whereas a module may or may not be enclosed in an object. As a class the module looks like (in ES5):

```js
// Sample.js
function Sample() {
  function gimmeString() {
  …
}

export default Sample;
```

And the test:

```js
// Sample-spec.js

describe( 'Sample', () => {
    describe( 'gimmeString()', () => {
        it( 'should … ', () => {
…
```

So the outer "describe" follows the classname and the filename. In a module, the exported object literal `{ init }` is inferred to have the name "sample" as the module is imported as that from because of the filename, but it's not as clear of a one-to-one name between the exported name and the "describe" block.

Probably not something to worry about to keep the example concise, but food for thought.